src/errors.js

```diff
@@ -0,0 +1,136 @@
+var stackTrace = require('stack-trace');
+var SourceMap = require('source-map');
+var colors = require('colors/safe');
+var fs = require('fs');
+
+function repeatString(string, count) {
+  return Array(count).join(string);
+}
+
+function pad(pad, str, padLeft) {
+  if (typeof str === 'undefined')
+    return pad;
+  if (padLeft) {
+    return (pad + str).slice(-pad.length);
+  } else {
+    return (str + pad).substring(0, pad.length);
+  }
+}
+
+function getArrow(length) {
+  return '  ' + repeatString('-', length + 1) + '^\n';
+}
+
+function getContextMessage(source, lineNumber, columnNumber) {
+  source = source.split('\n');
+  var lineDigits = ('' + (lineNumber + 1)).length + 1;
+  var padding = repeatString(' ', lineDigits);
+
+  var previousPrefix = colors.dim(pad(padding, (lineNumber - 1), true) + '| ');
+  var errorPrefix = colors.dim(pad(padding, lineNumber, true) + '| ');
+  var followingPrefix = colors.dim(pad(padding, (lineNumber + 1), true) + '| ');
+
+  var previousLine = source[lineNumber - 2] + '\n';
+  var errorLine = colors.bold(source[lineNumber - 1]) + '\n';
+  var followingLine = source[lineNumber] + '\n';
+
+  previousLine = previousLine.trim().slice(0, 2) === '//' ? colors.dim(previousLine) : previousLine;
+  followingLine = followingLine.trim().slice(0, 2) === '//' ? colors.dim(followingLine) : followingLine;
+
+  var previousTotal = previousLine === 'undefined\n' ? '' : previousPrefix + previousLine;
+  var errorTotal = errorPrefix + errorLine;
+  var followingTotal = followingLine === 'undefined\n' ? '' : followingPrefix + followingLine;
+
+  return previousTotal + errorTotal + (padding + getArrow(columnNumber)) + followingTotal;
+}
+
+// Note that this is Node specific.
+function showFriendlyError(error) {
+  if (!(error instanceof Error)) {
+    // Probably a string from `throw message`.
+    console.log(error);
+    return;
+  }
+
+  var pos = getErrorPosition(error);
+  var src = getSrc(error, pos);
+
+  if (src) {
+    writeFriendlyError(error, pos, src);
+  } else {
+    console.log('\nFailed to generate friendly error message. Original error was:\n');
+    console.log(error.message + '\n');
+  }
+
+  if (error.stack) {
+    console.log('Stack trace:');
+    console.log(error.stack.slice(error.stack.indexOf('\n') + 1));
+  }
+}
+
+function getSrc(error, pos) {
+
+  // `error.sourceMaps` will contain one or more source maps if the
+  // error occurred while evaluating a webppl program.
+
+  if (pos.sourceMapped) {
+    return getSrcFromMap(error.sourceMaps[0], pos.fileName);
+  } else {
+    try {
+      return fs.readFileSync(pos.fileName, 'utf8');
+    } catch (e) {
+      // This can occur if the original location gets lost during
+      // compilation.
+      return null;
+    }
+  }
+}
+
+function getSrcFromMap(sourceMap, fileName) {
+  var mapConsumer = new SourceMap.SourceMapConsumer(sourceMap);
+  return mapConsumer.sourceContentFor(fileName);
+}
+
+function writeFriendlyError(error, pos, src) {
```

Comment:
This sounds very similar to `showFriendlyError`. Give it a more distinctive name, or inline it?